src/trainers/KNNTrainer.js

```diff
@@ -14,96 +14,98 @@ import { logFirehoseMetric } from "../helpers/metrics";
 const KNN = require("ml-knn");
 
 export default class KNNTrainer {
-  calculatePossibleKValues() {
-    const state = this.store.getState();
-    let datasetSize = state.data.length;
-    let possibleKValues = [1, 3, 5, 7, 17, 31, 45, 61];
-    const heuristicK = Math.round(Math.sqrt(datasetSize));
-    possibleKValues.push(heuristicK);
-    const oneThird = Math.round(datasetSize / 3);
-    possibleKValues.push(oneThird);
-    return possibleKValues;
-  }
 
   startTraining(store) {
     this.store = store;
     const state = store.getState();
-    const datasetSize = state.data.length;
-    /*
-      We modify algorithm hyperparameters (k) based on dataset size and type of
-      machine learning in attempt to increase the liklihood of accurate
-      models that behave in ways consistent with the mental model presented in
-      the curriculum.
-    */
+
+    const trainedModel = this.getOptimalModelDetails(state);
+
+    this.storeTrainedModel(store, trainedModel);
+
+    const state2 = store.getState();
+
+    logFirehoseMetric("train-model", state2);
+
+    this.storeHistoricResult(store, state2);
+  }
+
+  /*
+    We modify algorithm hyperparameters (k) based on dataset size and type of
+    machine learning in attempt to increase the liklihood of accurate
+    models that behave in ways consistent with the mental model presented in
+    the curriculum. For large classification datasets we try a variety of K
+    values and select the one that yields the most accurate model.
+  */
+  getOptimalModelDetails(state) {
+    let optimalValues = {};
+    let bestModel = null;
+    let bestPredictedLabels = [];
+    let bestK = -1;
+    let bestAccuracy = -1;
+    const kValues = this.possibleKValues(state);
+    kValues.forEach(kValue => {
+      this.knn = new KNN(state.trainingExamples, state.trainingLabels, {
```

Comment:
I'm curious, is there a reason we need to assign the result to `this.knn` and then assign that to `var model`? Or could we just assign the result directly to `model`?